Library/Homebrew/software_spec.rb

```diff
@@ -402,22 +404,38 @@ def compatible_locations?
   # Does the {Bottle} this {BottleSpecification} belongs to need to be relocated?
   sig { returns(T::Boolean) }
   def skip_relocation?
-    cellar == :any_skip_relocation
+    legacy_cellar_or_cellar_for_tag == :any_skip_relocation
   end
 
   def tag?(tag)
     checksum_for(tag) ? true : false
   end
 
-  # Checksum methods in the DSL's bottle block optionally take
+  # BottleChecksum methods in the DSL's bottle block take
   # a Hash, which indicates the platform the checksum applies on.
-  Checksum::TYPES.each do |cksum|
-    define_method(cksum) do |val|
-      digest, tag = val.shift
-      collector[tag] = Checksum.new(cksum, digest)
+  # Example args:
+  # {"22948764d8f8d7be4870ff92dae64d986eb63a9150b219c20fff87d1a6aa93d6"=>
+  # {"tag"=>:mojave, "cellar"=>:any_skip_relocation}}
+  Checksum::TYPES.each do |checksum_type|
+    define_method(checksum_type) do |*args|
+      sha256_line_hash = args.first
+      checksum_hash = sha256_line_hash.select { |key, _value| key.length == 64 }
+      checksum_value = checksum_hash.keys.first
+      tag = checksum_hash[checksum_value]
+      # Legacy DSL: There is only a single cellar line for all the tags
+      # New DSL: get the cellar for the specific tag
+      cellar_for_tag = cellar.presence || sha256_line_hash["cellar"]
+      cellar_for_tag ||= Homebrew::DEFAULT_CELLAR
+      collector[tag] = BottleChecksum.new(checksum_value, tag: tag, cellar: cellar_for_tag)
     end
   end
 
+  def legacy_cellar_or_cellar_for_tag
+    # Legacy DSL: There is only a single cellar line for all the tags
+    # New DSL: get the cellar for the specific tag
+    cellar.presence || checksum_for(Utils::Bottles.tag).first.cellar
```

Comment:
I don't think `.first` is a good idea here.

Reply:
checksum_for(Utils::Bottles.tag) returns an array of 2 values: the BottleChecksum and a boolean if the tag exists. I guess this warrants another refactoring as this is probably not the best implementation.